src/structure: format and minimize trait bounds

spl opened the pull request:
There are two related changes here. Neither one involves a change to the function of the code. Both are localized to the `src/structure` directory to reduce the number of changes to review.

The first change is to run `rustfmt` over all of the source files in the directory. This is mostly a whitespace change.

The second is reduce the usage of trait bounds to only where they are needed. The main changes were:
* In general, the `struct`s did not need trait bounds. I removed all except one, which was necessary.
* There were a large number of `Clone` bounds that were not needed. I removed all except the necessary ones, which I attached to functions instead of the implementation.
* I reduced the number of `AsRef<[u8]>` bounds to make it clear which functions used it and which did not.

spl commented:
BTW, my motivation for submitting this is that I want to work on files in `src/structure`, and I think that will be easier if (a) the formatting is standardized (with `rustfmt`) and (b) the code is easier to work with (i.e. the reason for minimizing trait bounds). Of course, people can disagree (politely! 😉) on aspects of style, and I'm happy to discuss it.

matko commented:
Why do you consider trait bounds on functions better than having those bounds on the structures?

spl commented:
> Why do you consider trait bounds on functions better than having those bounds on the structures?

There are a couple of things here:

1. Consider the `LogArray` `struct` defined as:


   ```rust
   #[derive(Clone)]
   pub struct LogArray<M: AsRef<[u8]> + Clone> {
       len: u32,
       width: u8,
       len_bytes: usize,
       data: M
   }
   ```

   I can remove the trait bound on `M` (`AsRef<[u8]> + Clone`), and it has no affect the rest of the library. Given that fact, I see the trait bound as superfluous and contributing more to clutter than to clarity.

   The one exception to this in this PR is `AdjacencyListBuilder`, in which an associated type is used in the `struct` definition:

   ```rust
   pub struct AdjacencyListBuilder<F,W1,W2,W3>
   where F: 'static+FileLoad+FileStore,
         W1: 'static+AsyncWrite+Send,
         W2: 'static+AsyncWrite+Send,
         W3: 'static+AsyncWrite+Send {
       bitfile: F,
       bitarray: BitArrayFileBuilder<F::Write>,
       bitindex_blocks: W1,
       bitindex_sblocks: W2,
       nums: LogArrayFileBuilder<W3>,
       last_left: u64,
       last_right: u64
   }
   ```

   In this case, `F: FileStore` is needed to get `F::Write` in type of `bitarray`.

2. Most of the `struct` implementation blocks have trait bounds in the `impl` declaration. I've removed the ones in which the trait bounds do not apply to all functions in the `impl` block, pushing those trait bounds onto the inherent methods in which they are required.

   This is partially a matter of taste, but I think it can have a meaningful impact on the usability of the code. By having unrequired trait bounds on a function, the type of the function becomes misleading. And other functions that use that function may also then require possibly unnecessary trait bounds.

   In some cases, a `struct` may be constructed or used with different sets of trait bounds. So forcing unrequired trait bounds limits the use of the `struct`. I don't think this is the case for any `struct`s that I changed: for example, many `struct`s are not useful without `M: AsRef<[u8]>`; however, I think it's good practice to use minimal types and, therefore, only the required trait bounds.

   In my second commit, I added, for example, `where M: AsRef<[u8]>` to a number of functions. An alternative approach would be to have a second implementation section with those functions: `impl<M: AsRef<[u8]> Struct<M> { ... }`. I would be fine with this option.

matko commented:
While you're right that pushing the trait bounds down to a subset of the function allows more flexibility in using this library, I do have my doubts that this is more clear or better.

In my opinion, `AsRef<[u8]>` is not an incidental implementation detail of just some functions, but part of the identity of these structs. Their whole point is that they are some kind of interpretation of an underlying byte array. The trait bound documents this. There is no intended use case of these structs where you would not have some underlying byte storage.

If you have an intended use case for instantiating these structs without an underlying byte array, please let me know, but I don't want to apply this change just cause it is better in an abstract programming theory kind of way. Sorry.

Regarding rustfmt, I feel bad about not using it yet. I'll reformat all library code soon and intend to start enforcing proper formatting through a build check.

spl commented:
> While you're right that pushing the trait bounds down to a subset of the function allows more flexibility in using this library, I do have my doubts that this is more clear or better.

Indeed, that is one consequence of what I am suggesting in general. However, that is not my motivation here, because I see that it doesn't apply.

> In my opinion, `AsRef<[u8]>` is not an incidental implementation detail of just some functions, but part of the identity of these structs. Their whole point is that they are some kind of interpretation of an underlying byte array. The trait bound documents this. There is no intended use case of these structs where you would not have some underlying byte storage.

Fair enough.

> If you have an intended use case for instantiating these structs without an underlying byte array, please let me know, but I don't want to apply this change just cause it is better in an abstract programming theory kind of way. Sorry.

While I think my change is not better in an abstract way but in a very practical way (since it documents the minimal type of each function), I grant you that, as a collection of functions for a given `struct`, they are not useful without the implementation-level `AsRef<[u8]>`. I don't have any other use cases.

spl commented:
> Regarding rustfmt, I feel bad about not using it yet. I'll reformat all library code soon and intend to start enforcing proper formatting through a build check.

Thank you! I look forward to that. Should I wait for that before submitting any more PRs?

spl commented:
In the meantime, I'll close this, since you have no intention of merging it.

matko commented:
I pushed a full reformatting just now, so feel free to submit more PRs.
